### ISSUE TITLE
build error

### ISSUE THREAD

**Issue by sparber** [kaldi-asr/kaldi#735]
I was trying to build kaldi from latest master but it seems to be broken.

I'm getting undefined reference errors when trying to compile src/online2/online-nnet3-decoding.cc:

> online-nnet3-decoding.cc:(.text+0x10e): undefined reference to `kaldi::nnet3::DecodableNnet3SimpleOnline::DecodableNnet3SimpleOnline(kaldi::nnet3::AmNnetSimple const&, kaldi::TransitionModel const&, kaldi::nnet3::DecodableNnet3OnlineOptions const&, kaldi::OnlineFeatureInterface*)'
> kaldi-online2.a(online-nnet3-decoding.o): In function `kaldi::nnet3::DecodableNnet3SimpleOnline::~DecodableNnet3SimpleOnline()':
> online-nnet3-decoding.cc:(.text._ZN5kaldi5nnet326DecodableNnet3SimpleOnlineD2Ev[_ZN5kaldi5nnet326DecodableNnet3SimpleOnlineD5Ev]+0x3): undefined reference to `vtable for kaldi::nnet3::DecodableNnet3SimpleOnline'
> online-nnet3-decoding.cc:(.text._ZN5kaldi5nnet326DecodableNnet3SimpleOnlineD2Ev[_ZN5kaldi5nnet326DecodableNnet3SimpleOnlineD5Ev]+0x33): undefined reference to `kaldi::nnet3::CachingOptimizingCompiler::~CachingOptimizingCompiler()'
> kaldi-online2.a(online-nnet3-decoding.o): In function `kaldi::nnet3::DecodableNnet3SimpleOnline::~DecodableNnet3SimpleOnline()':
> online-nnet3-decoding.cc:(.text._ZN5kaldi5nnet326DecodableNnet3SimpleOnlineD0Ev[_ZN5kaldi5nnet326DecodableNnet3SimpleOnlineD0Ev]+0x3): undefined reference to `vtable for kaldi::nnet3::DecodableNnet3SimpleOnline'
> online-nnet3-decoding.cc:(.text._ZN5kaldi5nnet326DecodableNnet3SimpleOnlineD0Ev[_ZN5kaldi5nnet326DecodableNnet3SimpleOnlineD0Ev]+0x33): undefined reference to `kaldi::nnet3::CachingOptimizingCompiler::~CachingOptimizingCompiler()'
> collect2: error: ld returned 1 exit status
> 

I think online-nnet3-decoding.cc shouldn't be located in the online2 folder. I removed `online-nnet3-decoding.o` from online2/Makefile and also `online2-wav-nnet3-latgen-faster` and `../nnet3/kaldi-nnet3.a` from online2bin/Makefile

With these changes I was able to successfully build it and everything seems to be working.

**Comment by danpovey** [comment]
I guess you were using a mac-- I was able to reproduce this problem on my
mac but not on linux.
While messing around with it the problem mysteriously disappeared at some
point and I can't replicate it any more.  I checked in some extremely minor
changes to Makefiles and some whitespace changes to code, but I checked and
reverting these did not seem to make the problem reappear.  So I think it
was one of those temporary 'make' problems that can be resolved by doing
'make clean' and/or 'make depend' and 'make'.  Probably just 'make clean'
in online2 and online2bin and nnet3, then 'make depend' and then re-making
would work.

Regarding the directory-- I thought so at first, but I think we agreed in
the end that the 'online2' is a version number of the online decoding which
doesn't inherently relate to nnet2.

Dan


On Thu, Apr 28, 2016 at 7:55 AM, Alexander Sparber <notifications@github.com
